src/main/java/net/fabricmc/loom/build/MixinRefmapHelper.java

```diff
@@ -43,29 +51,72 @@
-			String refmapName = container.getRefmapName();
-
-			return ZipUtil.transformEntries(output, mixinJsonNames.map(f -> new ZipEntryTransformerEntry(f, new StringZipEntryTransformer("UTF-8") {
-				@Override
-				protected String transform(ZipEntry zipEntry, String input) {
-					JsonObject json = LoomGradlePlugin.GSON.fromJson(input, JsonObject.class);
-
-					if (!json.has("refmap")) {
-						json.addProperty("refmap", refmapName);
+		try {
+			MixinApExtension mixin = LoomGradleExtension.get(project).getMixin();
+			File output = outputPath.toFile();
+
+			Collection<String> allMixinConfigs = getMixinConfigurationFiles(readFabricModJson(output));
+
+			return mixin.getMixinSourceSetsStream().map(sourceSet -> {
+				MixinApExtension.MixinInformationContainer container = Objects.requireNonNull(
+						MixinApExtension.getMixinInformationContainer(sourceSet)
+				);
+
+				Stream<String> mixinConfigs = sourceSet.getResources()
+						.matching(container.mixinConfigPattern())
+						.getFiles()
+						.stream()
+						.map(File::getName)
+						.filter(allMixinConfigs::contains);
+
+				String refmapName = container.refmapNameProvider().get();
+
+				return ZipUtil.transformEntries(output, mixinConfigs.map(f -> new ZipEntryTransformerEntry(f, new StringZipEntryTransformer("UTF-8") {
+					@Override
+					protected String transform(ZipEntry zipEntry, String input) {
+						JsonObject json = LoomGradlePlugin.GSON.fromJson(input, JsonObject.class);
+
+						if (!json.has("refmap")) {
+							json.addProperty("refmap", refmapName);
+						}
+
+						return LoomGradlePlugin.GSON.toJson(json);
 					}
+				})).toArray(ZipEntryTransformerEntry[]::new));
+			}).reduce(false, Boolean::logicalOr);
+		} catch (Exception e) {
+			project.getLogger().error(e.getMessage());
+			return false;
+		}
+	}
+
+	@NotNull
+	private static JsonObject readFabricModJson(File output) {
+		try (ZipFile zip = new ZipFile(output)) {
+			ZipEntry entry = zip.getEntry(FABRIC_MOD_JSON);
 
-					return LoomGradlePlugin.GSON.toJson(json);
-				}
-			})).toArray(ZipEntryTransformerEntry[]::new));
-		}).reduce(false, Boolean::logicalOr);
+			try (InputStreamReader reader = new InputStreamReader(zip.getInputStream(entry))) {
+				return LoomGradlePlugin.GSON.fromJson(reader, JsonObject.class);
+			}
+		} catch (IOException e) {
+			throw new RuntimeException("Cannot read file fabric.mod.json in the output jar.", e);
+		}
+	}
+
+	@NotNull
+	private static Collection<String> getMixinConfigurationFiles(JsonObject fabricModJson) {
+		return StreamSupport.stream(fabricModJson.getAsJsonArray("mixins").spliterator(), false)
+				.map(e -> {
+					if (e instanceof JsonPrimitive str) {
+						return str.getAsString();
+					} else if (e instanceof JsonObject obj) {
+						return obj.get("config").getAsString();
+					} else {
+						throw new RuntimeException("Incorrect fabric.mod.json format");
+					}
+				}).collect(Collectors.toList());
```

Comment:
```suggestion
				}).collect(Collectors.toSet());
```
Might as well use a set here since duplicates don't matter (though I don't know how loader behaves with them)